bpf: always track egress gateway connections

jibi opened the pull request:
In order to reclaim orphan entries in the NAT table the agent uses the connection tracking state: if a NAT entry is not backed by a connection, it should be considered orphaned and can be reclaimed.

Currently the SNAT logic assumes that all connections that needs to be translated are already tracked. The only exception for this is connections originating from the local host, which are tracked by snat_v4_track_local().

In practice egress gateway is another case that requires special attention: in fact, all connections tunneled from a different node to a gateway node are currently not tracked. This means that whenever the CT garbage collector kicks in, all NAT entries for egress gateway will be deleted, potentially leading to connections breakage.

This commit fix this by making sure we always track these connections.

While at it, also rename snat_v4_track_local to snat_v4_track_connection.

Towards: #21346

julianwiedmann reviewed (commented):
High-level comments:

> Currently the SNAT logic assumes that all connections that needs to be translated are already tracked. The only exception for this is connections originating from the local host, which are tracked by snat_v4_track_local().

The expectation is that `from-container` would create the CT entry for any other connection, correct?

> In practice egress gateway is another case that requires special attention: in fact, all connections tunneled from a different node to a gateway node are currently not tracked. This means that whenever the CT garbage collector kicks in, all NAT entries for egress gateway will be deleted, potentially leading to connections breakage.

Any chance of adding a test for this scenario? If we have a mechanism to trigger GC, I suppose we could even just check the NAT table before/after.

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -1276,7 +1278,7 @@ static __always_inline int nodeport_nat_ipv4_fwd(struct __ctx_buff *ctx)
 	int ret = CTX_ACT_OK;
 
-	if (snat_v4_needed(ctx, &target.addr, &from_endpoint))
+	if (snat_v4_needed(ctx, &target.addr, &from_endpoint, &target.egress_gateway))
```
Can we pass the whole `target` as parameter instead, and let `snat_v4_needed()` adjust `.addr` and `.egress_gateway` as needed?

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -275,6 +276,14 @@ static __always_inline int snat_v4_track_local(struct __ctx_buff *ctx,
+		 * related to a remote endpoint (if the endpoint is local then
+		 * the connection is already tracked).
+		 */
+		else if (target->egress_gateway && !__lookup_ip4_endpoint(tuple->saddr))
```
There's a `__lookup_ip4_endpoint(ip4->saddr)` in `snat_v4_needed()` already, which I believe is the same thing...?

Can we make use of this in `snat_v4_needed()` to only set `.egress_gateway` for non-local connections? And then just consume the flag here? I'd prefer to keep the _details_ of EgressGW limited to as few places as possible.

jibi replied:
> I'd prefer to keep the details of EgressGW limited to as few places as possible.

agree with the general principle but in this case I'd prefer to keep the check explicit (even if it costs a second lookup) and not mix responsibilities (i.e. making `snat_v4_needed` responsible for figuring out if that connection should be tracked)

julianwiedmann replied:
If it feels like too much of a layering violation, fair enough. Although I don't see much difference between indicating `this is an EgressGW connection` and `this is a non-local EgressGW connection`. The NAT code still controls in what order it makes its actual decision, `snat_v4_needed()` just provides more precise information.

pchaigno replied:
I think I agree with Julian here. `snat_v4_needed` is already responsible for much more than checking if SNAT is needed. Maybe we should simply rename it to `snat_v4_preparation` or similar and have it collect the full state necessary for SNAT (saddr, whether it's egress gateway, AND whether it's local) and not half of it.

jibi replied:
Fair enough, let's go for the more generic prepare state/apply state then, might be useful in the future as well

julianwiedmann reviewed (changes requested):
Nice one @jibi, thank you! Reasoning is sound, just some fine-tuning aspects I'd like to clarify.

christarazi commented on test/k8s/egress.go:
```diff
@@ -76,8 +77,23 @@ var _ = SkipDescribeIf(func() bool {
+	// ctEntriesOnGatewayNode returns the number of CT entries on the
+	// designated gateway node (k8s2) matching destination IP and port
+	ctEntriesOnGatewayNode := func(dstIP, dstPort string) int {
+		ciliumPodK8s2, err := kubectl.GetCiliumPodOnNode(helpers.K8s2)
```
Maybe it makes sense to not have the function assume which is the gateway node, but rather leave it up to the caller to have that context?

pchaigno commented on test/k8s/egress.go:
```diff
@@ -76,8 +77,23 @@ var _ = SkipDescribeIf(func() bool {
+		ciliumPod, err := kubectl.GetCiliumPodOnNode(node)
+		Expect(err).Should(BeNil(), "Unable to determine cilium pod on node %s", node)
+
+		cmd := fmt.Sprintf("cilium bpf ct list global | grep '\\-> %s:%s' | wc -l", dstIP, dstPort)
```
Worth making sure the port isn't matching on more than intended?
```
grep '\\-> %s:%s\\b'
```
We've had such bugs in the past with IP addresses and it caused a long-lived flake.

jibi replied:
Right, nice catch :boom:

pchaigno reviewed (commented):
Thanks for the nice explanation and the test improvement!

jibi committed:
bpf: nat: always track egress gateway connections

In order to reclaim orphan entries in the NAT table the agent uses the
connection tracking state: if a NAT entry is not backed by a connection,
it should be considered orphaned and can be reclaimed.

Currently the SNAT logic assumes that all connections that needs to be
translated are already tracked. The only exception for this is
connections originating from the local host, which are tracked by
snat_v4_track_local().

In practice egress gateway is another case that requires special
attention: in fact, all connections tunneled from a different node to a
gateway node are currently not tracked. This means that whenever the CT
garbage collector kicks in, all NAT entries for egress gateway will be
deleted, potentially leading to connections breakage.

This commit fix this by making sure we always track these connections.

While at it, also rename snat_v4_track_local() to
snat_v4_track_connection().

Towards: #21346
Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
bpf: nat: move snat_v{4,6}_needed into lib/nat.h

No functional changes introduced.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
bpf: nat: fix from_endpoint assignment

In snat_v4_needed(), whenever NAT is required because of an egress
gateway policy, we unconditionally set the from_endpoint parameter to
true, as egress gateway traffic is always originating from an endpoint.

This is wrong as that parameter is supposed to track whether traffic is
coming from a _local_ endpoint, and so that variable should be set to true
only if the __lookup_ip4_endpoint() lookup is successful.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
bpf: nat: consolidate snat_v4_needed() state into ipv4_nat_target

Move the from_endpoint parameter of snat_v4_needed() into
ipv4_nat_target, and rename the function to snat_v4_prepare_state().

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
bpf: nat: use target->from_local_endpoint in snat_v4_track_connection()

Use the information stored in the SNAT target state to determine if the
packet is coming from a local endpoint, saving one map lookup.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

pchaigno commented on bpf/lib/nat.h:
```diff
@@ -257,11 +258,11 @@ static __always_inline int snat_v4_new_mapping(struct __ctx_buff *ctx,
-					       const struct ipv4_nat_entry *state,
-					       enum nat_dir dir, __u32 off,
-					       const struct ipv4_nat_target *target)
+static __always_inline int snat_v4_track_connection(struct __ctx_buff *ctx,
```
Should we also rename `snat_v6_track_local`, if only to make sure the two code paths don't diverge too much despite egress gateway still being IPv4-only?

jibi replied:
Yep, good shout :+1:

pchaigno commented on bpf/lib/nat.h:
```diff
@@ -558,22 +560,20 @@ static __always_inline void snat_v4_init_tuple(const struct iphdr *ip4,
-static __always_inline bool snat_v4_needed(struct __ctx_buff *ctx,
-					   struct ipv4_nat_target *target,
-					   bool *from_endpoint __maybe_unused)
+static __always_inline bool snat_v4_prepare_state(struct __ctx_buff *ctx,
```
It pains me to suggest this because my IPv6 masquerading PR rebase will hurt, but: should we also rename the IPv6 function accordingly?

jibi replied:
I looked at renaming this function as well, but it's not really preparing any state, it's just running a couple of checks on DSR and source address :thinking: so fine for me to rename it for consistency, but the name may be a bit misleading (at least for now)

pchaigno replied:
Agree. I don't feel strongly either way.

jibi replied:
Let's leave it as it is then, the `/* See snat_v4_prepare_state(). */` comment should be already enough for now to cross reference it to the v4 implementation

pchaigno reviewed (approved):
Thanks! Nice commit history btw :tada:

jibi committed:
bpf: nat: rename snat_v6_track_local to snat_v6_track_connection

For consistency with its v4 counterpart.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>
